File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.cpp
class Solution {
public:
    int longestIdealString(string s, int k) {
        vector<int> dp(27, 0);
        int n = s.length();

        for (int i = n - 1; i >= 0; i--) {
            int cc = s[i] - 'a';
            int maxi = INT_MIN;

            int left = max(cc - k, 0);
            int right = min(cc + k, 26);

            for (int j = left; j <= right; j++) {
                maxi = max(maxi, dp[j]);
            }

            dp[cc] = maxi + 1;
        }

        return *max_element(dp.begin(), dp.end());
    }
};

File: 2370-longest-ideal-subsequence/2370-longest-ideal-subsequence.js
var longestIdealString = function(s, k) {
    let dp = new Array(27).fill(0);
    let n = s.length;

    for (let i = n - 1; i >= 0; i--) {
        let cc = s[i];
        let ind = cc.charCodeAt(0) - 'a'.charCodeAt(0);
        let maxi = -Infinity;

        let left = Math.max(ind - k, 0);
        let right = Math.min(ind + k, 26);

        for (let j = left; j <= right; j++) {
            maxi = Math.max(maxi, dp[j]);
        }

        dp[ind] = maxi + 1;
    }

    return Math.max(...dp);
};